data/posts/middleware.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Middleware in React Router
+summary: "Middleware is now stable in React Router!"
+featured: false
+date: 2025-09-12
+image: /blog-images/headers/fog-of-war.png
+ogImage: /blog-images/headers/fog-of-war.png
+imageAlt: React Router Middleware
+imageDisableOverlay: false
+authors:
+  - Matt Brophy
+---
+
+The long-awaited middleware feature is now stable in React Router [7.9.0][release-notes-7-9-0] (behind a `future.v8_middleware` flag)! It's been a long road to get here, but we wanted to make sure we got this one right. We think it was worth the wait.
+
+## Remix v1/v2 patterns
+
+React Router has always embraced nested routes. Remix layered on a compelling SSR story which included built-in [data loading primitives][data-loading]. To keep your application fast, route loaders ran in parallel for all matched routes for a given URL. This was crucial for performance, but meant that you couldn't do any sequential logic in your loaders, nor could you short-circuit out from a parent route. We received lots of questions and saw lots of incorrect attempts to do both of these in early Remix apps.
+
+Often, folks wanted to access parent loader data from a child loader - but this wasn't possible because they run in parallel:
+
+```tsx
+// app/root.tsx
+export async function loader({ request }) {
+  let user = await getUser(request);
+  return { user };
+}
+
+// app/routes/_index.tsx
+export async function loader({ request }) {
+  // ❌ No way to access `user` here
+}
+```
+
+Other times, folks would incorrectly assume a redirect from a parent loader would prevent the child `loader` from running - but this was also not the case:
+
+```tsx
+// app/routes/_auth.tsx
+export async function loader({ request }) {
+  let user = await getUser(request);
+  if (!user) {
+    throw redirect("/login");
+  }
+}
+
+// app/routes/_auth.profile.tsx
+export async function loader({ request }) {
+  // ❌ This code still runs even if there's no logged-in user
+  let data = await getUserData(request);
+  return data;
+}
+```
+
+These were easy mistakes to make because nested routes could lead you to assume this would work - especially coming from the "load data in your component" patterns that were popular in React Router SPA's prior to Remix. We documented these behaviors clearly, but we always knew that it was a gap in our API.
```

Comment:
```suggestion
These were easy mistakes to make because nested routes could lead you to assume this would work - especially coming from the "load data in your component" patterns that were popular in React SPA's prior to Remix. We documented these behaviors clearly, but we always knew that it was a gap in our API.
```